Fix protocol violation: Respond to SMFIC_ABORT with no response, not SMFIS_CONTINUE

igor-winograd opened the pull request:
Hi @girstenbrei,

I hope you are doing well.

First, thank you again for this excellent library — it's been a great tool to work with.

During the integration of miltr with Postfix in a scenario where a single connection processes multiple emails, I encountered a persistent protocol error. Postfix would abruptly terminate the connection, logging the following error:

unexpected filter response SMFIR_INSHEADER after event SMFIC_MAIL

After analyzing the issue with tcpdump and studying the Milter protocol specification (a copy of which I've included in the root folder for easy future reference), I pinpointed the root cause.

The protocol explicitly states for the SMFIC_ABORT command:
```
  'A'  SMFIC_ABORT  Abort current filter checks
                   Expected response:  NONE
```
However, the library currently responds to an SMFIC_ABORT command with an SMFIS_CONTINUE action. 
This is a protocol violation. While Postfix might occasionally tolerate this, it eventually leads to a critical desynchronization of the protocol state machine. 

The MTA, having sent an abort signal, expects no response and resets the transaction state. 
The milter, however, by responding, creates a mismatch that culminates in the observed error when the milter attempts to modify headers (SMFIR_INSHEADER) for a non-existent transaction.

The fix:
This PR updates the handler for the SMFIC_ABORT command to send no response. This ensures proper state synchronization between the milter and the MTA across multiple messages within a single connection.

I have tested this fix, and it completely resolves the connection termination issue, enabling stable and reliable processing of multiple emails per connection.

This issue is currently a blocker for my use case, and I would be immensely grateful if you could review and consider merging this correction.

Thank you for your time and for your work on this project.

Best regards,
Igor

igor-winograd committed:
fix abort response

igor-winograd committed:
fix fmt

igor-winograd committed:
fix missing backticks in doc

girstenbrei commented:
Hi @igor-winograd , thank you very much for the PR! I poked around a little and apparently we do have an issue with connection reuse, yes, good catch!

But I tried to reproduce the problem as well as the fix and I seem to cannot do both. Could you share how you produced the problem? And you mentioned you tested it works now, could you share that test as well?

girstenbrei commented:
OK, some more tinkering later, I think I got it - you are correct, responding to postfix is not the correct thing to do.

The server needs to run with `Server::new(&mut milter, false, 2_usize.pow(16));`, notably: `quit_on_abort: false`. Especially `Server::default_postfix()` does not do the intuitive thing.

So, to sort this out I'd like to do combine the following:

- Your change
- I'd like to not include the milter spec TBH, I'd like to link to it. This repo is not an authority on the spec, that should IMHO live somewhere else.
- Add a test to ensure we have the desired behavior for multiple mails over the same connection
- Change the `Server::default_postfix` to match expectations.

Does that sound good to you? Do you want continue with this plan in this PR or should I pick up your commit and go from there?

Have a nice day!
Christoph

igor-winograd commented:
Hi Christoph,

That sounds like an excellent and comprehensive plan. I fully agree with your approach.

I am happy to continue implementing these steps in this PR:
1. Keep the core change to not respond to `SMFIC_ABORT`.
2. Remove the spec file.
3. Work on adding a test for multi-email connection stability.
4. Update `Server::default_postfix()` to set `quit_on_abort: false` by default.

This will definitely make the library more intuitive and robust. I'll push the updates shortly.

Thank you for the guidance!

Best regards,
Igor

igor-winograd committed:
add tests

igor-winograd commented:
Hi @girstenbrei,

I've implemented the changes we discussed.

Please take a look when you get a chance - I'm happy to iterate based on your feedback.

Cheers,
Igor

girstenbrei commented:
Hi, @igor-winograd ,

thanks for the test, I have a few suggestions here:

- Could you remove the milter-protocol.txt and link to the original specification in the docs? I don't want to suggest that this spec is owned by this repository and up to change
- The way you implemented `MilterMock` generates a lot of code barely using any of it in the `many_mails` (2 out of 15 methods). The [body modification tewst](https://github.com/girstenbrei/miltr/blob/main/server/tests/modifications/body.rs) is an example of a milter doing something, even simpler - the [Action milter](https://github.com/girstenbrei/miltr/blob/main/server/tests/utils/action_milter.rs) is a shortcut to just call an action at EOB. I think (correct me if wrong) that you implemented it this way, because `TestCase::send_mail` does not give you the option to re-use the connection ATM. But I think this is the simpler way to implement this. Add two pub methods, maybe `TestCase::setup_transport`  and `TestCase::send_mail_with_transport`. Keep `TestCase::send_mail`, but change it's body to use the new methods. Then all existing test cases don't need any change. Your new test case can use those two new methods to create a transport and send multiple mails.
- I think the testcase does not actually test the behavior that you would like to test - that only one connection is created towards the milter and then re-used. The asserts currently test that "nothing hugely wrong happened" and "something happened", which is good, but the positive "the thing we want happened" is missing. I think one way to do it is to expose a counter from `run_milter` and store it in `TestCase`, having a getter for it. Then the test case can run and at the end assert the number of created connections.
- I don't think there is a reason to [sleep in the test](https://github.com/girstenbrei/miltr/pull/22/files#diff-e8a7880a9eac46366b7aca06902ffb41d431dae615a5deebd006008928d40c8cR21-R26). It does not achieve anything: No new behaviors are covered, no outcome is computed. I think you can just remove it. Additionally, could you introduce a constant number that is [used as the end in the for loop](https://github.com/girstenbrei/miltr/pull/22/files#diff-e8a7880a9eac46366b7aca06902ffb41d431dae615a5deebd006008928d40c8cR48) and then in [the asserts about called fns](https://github.com/girstenbrei/miltr/pull/22/files#diff-e8a7880a9eac46366b7aca06902ffb41d431dae615a5deebd006008928d40c8cR68) ? I think this would enhance clarity about how many times we expect things to happen.
- Sorry to be nit-picky here, but the naming of the new module `integration` suggests, the other tests are not integration tests. But all of them are. I'm not sure what to name them, maybe "connection" as we might want more tests that test connection behavior in the future.

Thank you very much for your work, let me know if something is wrong with my feedback. Maybe a last comment - especially the `milter_mock.rs` reads AI. Relatively large amount of code, only a very small amount used. And, I accidentally know, that this callback-style is how other milter implementations do it. I'm not saying, don't use AI, just be careful if you do. Maybe I'm also wrong :smile: 

Greets,
Christoph

igor-winograd committed:
remove milter-protocol.txt

Commit:
fix test/remove MilterMock

igor-winograd commented:
Hi @girstenbrei ,

Thank you very much for the detailed and constructive analysis of the code!

I made edits based on your suggestions:

- I forgot to remove the `milter-protocol.txt` file, my apologies for that.

- You are right, `MilterMock` does look large and unused. I've gotten rid of it.

- I've tried to address the feedback regarding connection handling by implementing the additional methods while keeping the existing `send_mail` function.

- I've added a connection counter in `run_milter`.

- I've added constants for the number of emails to send and for the wait time in `Mail` handler.

- I completely agree with the point about the naming. "Integration" doesn't seem fitting. I've changed it to "Connection".

- However, I wasn't able to reuse the `AsyncSmtpTransport<Tokio1Executor>`. Its connection pool can recreate connections, which caused the `many_mail_send_test` to behave unpredictably.

- I couldn't remove the sleep in `Mail` handler. Without it, the original protocol error only reproduces around the ~70th-100th email. It seems that if the mail processor's response is delayed, Postfix starts processing the response from the previous EOB and throws an error.


Best regards,
Igor

girstenbrei commented:
Very nice @igor-winograd !

- Nice cleanup
- Good catch with the `AsyncSmtpTransport`, `AsyncSmtpConnection` seems like a good solution
- Thx for the const and docs regarding `MAIL_COMMAND_SLEEP_DURATION`
- Nice checking for the sleep problem. Maybe I'll investigate this later...

Last, hopefully simple nitpicks:

- Could you rename the test `many_mail_send_test` to `test_many_mail_send` to match other test names?
- `TestCase<S>::create_smtp_transport()` needs to be in the other impl block - `TestCase<RunningState>::create_smtp_transport()`. Otherwise you can `testcase.stop()` and then call `testcase.create_smtp_transport()`.
- I don't think we need a `ConnectCounter`. `TestCase.connect_count` can just be an `Arc<AtomicU64>`. Also implements default, no need for locking & unwraps.
- Do we need the `AddHeader` modifications in the `end_of_body` for this test to work? If yes, could you doc why? Otherwise removing them to avoid confusion I think is best.

Thank you again, and this should be the last nitpicks!

Have a nice day,
Chirstoph

Commit:
remove ConnectCounter/rename test/move create_smtp_transport

igor-winograd commented:
Hi @girstenbrei!
Thank you for your time and for the detailed review of my PR. Your suggestions are very useful and help make the code cleaner and more understandable.

I have renamed the test, moved the `create_smtp_transport` logic, and changed the `ConnectCounter` struct to use `Arc<AtomicUsize>`.

Regarding the `EOM` modifications and the `Abort` handler:
When we send `Action::Continue` in the `Abort` handler, Postfix sometimes incorrectly treats it as a response to the next command (e.g., as the final action for modification responses). Then, in the next `Mail` handler, Postfix processes the first response as a modification and throws an error.

You can reproduce this by adding the following code to process `Abort` in `lib.rs::handle_connection`:

```rust
let c: Action = miltr_common::actions::Continue.into();  
framed.send(&c.into()).await?;
```
After this, if you run the test, the full error message in the Postfix log will be something like:
`unexpected filter response SMFIR_INSHEADER after event SMFIC_MAIL`

Hopefully this explains why the test requires modification response to `EOM`.

If you have any more suggestions, I will correct everything without any problem.

Have a nice day!
Igor

girstenbrei commented:
Superb @igor-winograd  , and I think I finally got it:

When not re-using the connection, the erroneously written action after the abort was just lost in the socket buffer. The connection closed, the buffer had some data in it, but Postfix just never reads it. TCP finishes, TCP buffer is just dropped.

But sending multiple mails via the same connection, we have written that additional response. Now Postfix does read the additional command, is now out of sync and the `ModificationResponse` is not as a response to EOB any more (the only place it is allowed). That's why Postfix complains now about `SMFIR_ADDHEADER` when it tries to read the `SMFIC_MAIL` response.

The only thing I could not reproduce is the timing issue requiring the `sleep(MAIL_COMMAND_SLEEP_DURATION)` in the mail command. For me, the above argument holds consistently without the sleep: If i add the additional action, it always fails. If I don't have it, it always works. Could  you share where your numbers of only being able to reproduce it every 70 - 100th mail come from?

We can merge this as-is, this just smells like something else is going on or we haven't considered another case, so just for my curiosity :smile: 

Cheers,
Christoph

Commit:
remove sleep from mail handler

igor-winograd commented:
Hi @girstenbrei!

I've reproduced the behavior you described by connecting to Postfix on the remote machine.

To get the test to fail consistently, I found it necessary to add a sleep in `Mail` handler. Without it, I had increase the number of emails sent to trigger the issue.

I also ran the test in a Docker environment on the same machine (without the additional wait) and was able to confirm the same results you experienced.

This leads me to believe that network latency can sometimes mask this underlying timing issue.

I have removed the additional wait from `Mail` handler, as it doesn't seem relevant for reproducing this in our test environment, where everything runs locally on the same machine.

Best regards,
Igor
